### PR TITLE
Feature/its 49

### PR DESCRIPTION
**Pull request by aalmansa-e2x**
This is the initial work needed to create initial category import file.

### CODE REVIEW

**Review by hamzalatif-e2x — APPROVED** [review]
Looks good to me, will leave for a bit for others to review

**Review by gina-frankel-e2x — APPROVED** [review]

File: src/bigcommerce/BigCommerceAPI.ts [hunk 1]
```diff
@@ -1,44 +1,42 @@
...
-        } catch (err) {
-            console.log(`Error creating category in BigCommerce: ${err}`);
-            return err;
-        }
-
+                    "X-Auth-Token": this.authToken,
+                },
+            }
+        );
+        return response.data.data as BCCategory;
```
Comment (gina-frankel-e2x) [inline src/bigcommerce/BigCommerceAPI.ts:38]: This is the BCCategory number, right?

File: src/cloud/storage/bucketUtils.ts [hunk 2]
```diff
@@ -1,11 +1,10 @@
 import {Storage} from "@google-cloud/storage";
```
Comment (gina-frankel-e2x) [inline src/cloud/storage/bucketUtils.ts:1]: This file is the one that gets the csv file from google -cloud, right?

File: src/firestore/categoryRepository.ts [hunk 3]
```diff
@@ -0,0 +1,47 @@
+import * as admin from "firebase-admin"
+import {firestore} from "firebase-admin/lib/firestore";
+import Firestore = firestore.Firestore;
+import QuerySnapshot = firestore.QuerySnapshot;
+
+export type Category = {
+    id: string,
+    path: string,
+}
+
```
Comment (gina-frankel-e2x) [inline src/firestore/categoryRepository.ts:10]: Is this going to be the pattern for the types ? ie that they are not all grouped together in one class?

File: src/index.ts [hunk 4]
```diff
@@ -1,20 +1,20 @@
...
 
-export const goodbyeWorld: HttpFunction = (req, res) => {
-  util();
-  res.send('Goodbye, World!');
-};
-
-var firestoreClient = admin.initializeApp({ projectId: "test" }).firestore()
+export const categoryImportFile  = async (file: any, context: CloudFunctionsContext) => {
+  await handleCategoryImport(file, context);
+}
```
Comment (gina-frankel-e2x) [inline src/index.ts:15]: This function is related to importing the file from google cloud, right?

File: src/service/categoryService.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
+import {BCCategory, bigCommerceAPI, BigCommerceAPI} from "../bigcommerce/BigCommerceAPI";
+import {Category, categoryRepository} from "../firestore/categoryRepository";
+
+export class CategoryService {
+
+    async createCategoryPath(categoryPath: string): Promise<void> {
+
+        const categoryStack: string[] = new Array();
+
+        let category: Category | null = {id: "0", path: ""};
```
Comment (gina-frankel-e2x) [inline src/service/categoryService.ts:10]: What does the `null = {id: "0", path: ""}` mean?

File: src/service/categoryService.ts [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        let parentId: number = 0;
+
+        while (categoryPath != null && categoryPath.length > 0) {
+            category = await categoryRepository.findByPath(categoryPath);
+            if (category == null) {
+                const lastCategory = categoryPath.substr(categoryPath.lastIndexOf("/") + 1, categoryPath.length);
```
Comment (gina-frankel-e2x) [inline src/service/categoryService.ts:16]: Think this could be extracted as a method to make it clearer

File: src/service/categoryService.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+export class CategoryService {
+
+    async createCategoryPath(categoryPath: string): Promise<void> {
+
+        const categoryStack: string[] = new Array();
+
+        let category: Category | null = {id: "0", path: ""};
+        let parentId: number = 0;
+
+        while (categoryPath != null && categoryPath.length > 0) {
```
Comment (gina-frankel-e2x) [inline src/service/categoryService.ts:13]: This this could be extracted to make it easier to read

File: src/service/categoryService.ts (hunk 6 above)
Comment (gina-frankel-e2x) [inline src/service/categoryService.ts:16]: The argument could be extracted two variables to make it clearer `categoryPath.lastIndexOf("/") + 1, categoryPath.length)`

File: test/firestore/categoryRepository.spec.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+                            empty: true,
+                            docs: []
+                        })
+                    })
+                })
+            })
+        })
+    }
+});
+
```
Comment (gina-frankel-e2x) [inline test/firestore/categoryRepository.spec.ts:26]: is my understanding correct: 
so for this:
1.  there is an `initializeApp` function in firebase,
2. which then returns a value which has a `firestore` method.
3. The firestore method return value is mocked.
- The return value  of the firestone method is then mocked 
and so on .... until the get method 

The return values of the `get` include the properties `empty` and `docs`
- you are telling the mock that `empty `
- must return `false` and `docs` must return `docs: [{id: 1, data: jest.fn().mockReturnValue({path: "test-path"})}]`

File: test/firestore/categoryRepository.spec.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+describe('findByPath', () => {
+    // const firestoreMock = firestore as jest.MockedFunction<any>
+
+    it('should find category by path', async () => {
+        const categoryRepository = new CategoryRepositoryService();
+        const category = await categoryRepository.findByPath('test-path');
+        expect(category).toStrictEqual({id: 1, path: 'test-path'})
```
Comment (gina-frankel-e2x) [inline test/firestore/categoryRepository.spec.ts:33]: What happens if it not a strictEqual ?

File: src/firestore/categoryRepository.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    async findByPath(path: string): Promise<Category | null> {
+        try {
+            const documentReference = this.firestoreClient.collection('categories');
+            const snapshot: QuerySnapshot = await documentReference.where("path", "==", path).get();
+
+            if (snapshot.empty) {
+                return Promise.resolve(null);
+            }
+            return Promise.resolve({id: snapshot.docs[0].id, path: snapshot.docs[0].data().path});
```
Comment (gina-frankel-e2x) [inline src/firestore/categoryRepository.ts:33]: maybe the arguments could be extracted  to variables  
 `path: snapshot.docs[0].data().path`  has lots of properties which makes it a bit hard to read if you are not that aware of firestone

File: src/bigcommerce/BigCommerceAPI.ts (hunk 1 above)
Comment (aalmansa-e2x) [reply src/bigcommerce/BigCommerceAPI.ts:38]: It's the whole response data object from BC

File: src/firestore/categoryRepository.ts (hunk 3 above)
Comment (aalmansa-e2x) [reply src/firestore/categoryRepository.ts:10]: we will group them when there are more than 1

File: test/firestore/categoryRepository.spec.ts (hunk 8 above)
Comment (aalmansa-e2x) [reply test/firestore/categoryRepository.spec.ts:26]: Can't see what file this refers to, but it sounds like what is doing yes

**Comment by aalmansa-e2x** [comment]
I will make the code a bit more readable in the next iteration

**Pull request merged**